### PR TITLE
Repartitioner for communication pattern

### PR DESCRIPTION
**Pull request by greole**
This PR implements the first part of the repartitioner namely repartitioning of the communication pattern.

Repartitioning means that based on the parameter `ranks_per_gpu` a set of owning and non-owning ranks are created. Owning ranks take over all the matrix rows and communication of the non owning ranks. 

To repartition the communication patter the following steps are performed:
- Check all communication partners on every rank and decide whether they are on a remote rank or local after repartitioning
- All communication to and from local after repartitioning ranks can be discarded
- The remaining non-local communication needs to be transferred to new owner rank

### CODE REVIEW

File: include/OGL/Repartitioner.H [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    */
+    std::tuple<std::shared_ptr<SparsityPattern>,
+               std::shared_ptr<SparsityPattern>,
+               std::vector<std::pair<bool, label>>>
+    repartition_sparsity(
+        const ExecutorHandler &exec_handler,
+        std::shared_ptr<SparsityPattern> src_local_pattern,
+        std::shared_ptr<SparsityPattern> src_non_local_pattern) const;
+
+    /* @brief repartion a communication pattern
```
Comment (chihtaw) [inline include/OGL/Repartitioner.H]:
```suggestion
    /* @brief repartition a communication pattern
```

File: include/OGL/Repartitioner.H [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+     *  and a pair of locality information, where the first bool stores if the
+     *  interface performs a local communication and the second bool if the
+     *  the interface was originally from this rank
+     */
+    std::pair<SparsityPattern, std::vector<bool>> build_non_local_interfaces(
+        const ExecutorHandler &exec_handler, SparsityPattern &loc,
+        const SparsityPattern &non_loc) const;
+
+
+    /*
```
Comment (chihtaw) [inline include/OGL/Repartitioner.H]: Could you add some info about the input parameters in the comment?

File: include/OGL/Repartitioner.H [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    std::shared_ptr<const partition> get_orig_partition() const
+    {
+        return orig_partition_;
+    };
+
+
+    /* @brief given received interfaces this function sorts into local and non
+     *local
```
Comment (chihtaw) [inline include/OGL/Repartitioner.H]: It is a little confusing here.

According to the comment here, the function sorts a given set of interfaces into local and nonlocal interfaces.
But the function takes local and nonlocal sparsity patterns and returns a new non-local sparsity pattern.

Does the function actually do a sorting job for a given set of interfaces?

File: include/OGL/Repartitioner.H [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+     * */
+    bool reparts_to_local(const ExecutorHandler &exec_handler, label rank) const
+    {
+        return get_owner_rank(exec_handler) ==
+               compute_owner_rank(rank, ranks_per_gpu_);
+    };
+
+    label get_ranks_per_gpu() const { return ranks_per_gpu_; }
+
+    /* @brief computes the size after repartitioning
```
Comment (chihtaw) [inline include/OGL/Repartitioner.H]:
```suggestion
    /* @brief computes the size of the submatrix owned by a given rank after repartitioning
```

File: include/OGL/Repartitioner.H [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+{
+    std::vector<label> p(vec.size());
+    std::iota(p.begin(), p.end(), 0);
+    std::stable_sort(p.begin(), p.end(), [&](std::size_t i, std::size_t j) {
+        return compare(vec[i], vec[j]);
+    });
+    return p;
+}
+
+/** @class Collects functions to repartion communication patterns
```
Comment (chihtaw) [inline include/OGL/Repartitioner.H]:
```suggestion
/** @class Collects functions to repartition communication patterns
```

File: include/OGL/Repartitioner.H [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+     *  interface performs a local communication and the second bool if the
+     *  the interface was originally from this rank
+     */
+    std::pair<SparsityPattern, std::vector<bool>> build_non_local_interfaces(
+        const ExecutorHandler &exec_handler, SparsityPattern &loc,
+        const SparsityPattern &non_loc) const;
+
+
+    /*
+    ** returns the the new sparsity patterns (local and non-local) and a vector
```
Comment (chihtaw) [inline include/OGL/Repartitioner.H]:
```suggestion
    ** returns the new sparsity patterns (local and non-local) and a vector
```

File: include/OGL/Repartitioner.H (hunk 3 above)
Comment (greole) [reply include/OGL/Repartitioner.H]: Good catch, this comment is outdated. I'll fix it with the PR that implements the function.

**Pull request merged**